openslides/agenda/views.py

```diff
@@ -208,6 +216,137 @@ def get_queryset(self):
+        """
+        Special view endpoint to add users to the list of speakers or remove
+        them. Send POST {'user': <user_id>} to add a new speaker. Omit
+        data to add yourself. Send DELETE {'speaker': <speaker_id>} to remove
+        someone from the list of speakers. Omit data to remove yourself.
+
+        Checks also whether the requesting user can do this. He needs at
+        least the permissions 'agenda.can_see' (see
+        self.check_permission()). In case of adding himself the permission
+        'agenda.can_be_speaker' is required. In case of adding someone else
+        the permission 'agenda.can_manage' is required. In case of removing
+        someone else 'agenda.can_manage' is required. In case of removing
+        himself no other permission is required.
+        """
+        # Retrieve item.
+        item = self.get_object()
+
+        if request.method == 'POST':
+            # Retrieve user_id
+            user_id = request.data.get('user')
+
+            # Check permissions and other conditions. Get user instance.
+            if user_id is None:
+                # Add oneself
+                if not self.request.user.has_perm('agenda.can_be_speaker'):
+                    self.permission_denied(request)
+                if item.speaker_list_closed:
+                    raise ValidationError({'detail': _('The list of speakers is closed.')})
+                user = self.request.user
+            else:
+                # Add someone else.
+                if not self.request.user.has_perm('agenda.can_manage'):
+                    self.permission_denied(request)
+                try:
+                    user = get_user_model().objects.get(pk=int(user_id))
+                except (ValueError, get_user_model().DoesNotExist):
+                    raise ValidationError({'detail': _('User does not exist.')})
+
+            # Try to add the user. This ensurse that a user is not twice in the
+            # list of coming speakers.
+            try:
+                Speaker.objects.add(user, item)
+            except OpenSlidesError as e:
+                raise ValidationError({'detail': e})
+            message = _('User %s was successfully added to the list of speakers.') % user
+
+        else:
+            # request.method == 'DELETE'
+            # Retrieve speaker_id
+            speaker_id = request.data.get('speaker')
+
+            # Check permissions and other conditions. Get speaker instance.
+            if speaker_id is None:
+                # Remove oneself
+                queryset = Speaker.objects.filter(
+                    item=item, user=self.request.user).exclude(weight=None)
+                try:
+                    # We assume that there aren't multiple entries because this
+                    # is forbidden by the Manager's add method. We assume that
+                    # there is only one speaker instance or none.
+                    speaker = queryset.get()
```

Comment:
Then I can't catch the Speaker.DoesNotExist exception to validate whether the user is on the list of speakers or a speaker with the given id exists.